### PR TITLE
Version0 final

### PR DESCRIPTION
**Pull request by shutaoz**

### CODE REVIEW

File: src/main/java/Problem1.java [hunk 1]
```diff
@@ -2,9 +2,50 @@
...
+
+        while(mid <= high) {
+            //case 1 when 1 encounters another 1
+            if(nums[mid] == 1) ++mid;
+
+            //case 2 when 1 encounters a 0
+            else if(nums[mid] == 0){
+                temp = nums[low];
+                nums[low] = nums[mid];
+                nums[mid] = temp;
```
Comment (pdgetrf) [inline src/main/java/Problem1.java:26]: I see repeated logic here. refactor the swap into a private function. 

we prefer single source of truth.

File: src/main/java/Problem3/Book.java [hunk 2]
```diff
@@ -16,10 +16,18 @@ public Book(String title, String author) {
...
+
     }
 
     @Override
     public boolean equals(Object obj) {
         // homework
+        if(obj == null) return false;
+        else if(obj.getClass() != this.getClass())  return false;
+
+        return ((Book)obj).title.equals(this.title) && ((Book)obj).author.equals(this.author) && ((Book)obj).id.equals(this.id);
```
Comment (pdgetrf) [inline src/main/java/Problem3/Book.java:31]: wrong logic 

```
A movie equals another movie only when their id are the same. Same rule for books being equal.
```

File: src/main/java/Problem3/Movie.java [hunk 3]
```diff
@@ -10,14 +10,23 @@
...
+        this.id = anotherMovie.id;
     }
 
     @Override
     public boolean equals(Object obj) {
         // homework
+        if(obj == null) return false;
+        else if(obj.getClass() != this.getClass())  return false;
+
+        return ((Movie)obj).rating.equals(this.rating) && ((Movie)obj).title.equals(this.title) && ((Movie)obj).id.equals(this.id);
```
Comment (pdgetrf) [inline src/main/java/Problem3/Movie.java:30]: same here. wrong logic.

File: src/main/java/Problem3/MovieAction.java [hunk 4]
```diff
@@ -22,5 +24,8 @@ public int getLateFeeInDollar() {
     @Override
     public int calcLateFees(int numOfDaysPastDue) {
         // homework
+        if(numOfDaysPastDue <= 0) return 0;
+        if(numOfDaysPastDue < 5) return lateFeePerDayInDollar * numOfDaysPastDue;
+        else return 2 * lateFeePerDayInDollar * numOfDaysPastDue;
```
Comment (pdgetrf) [inline src/main/java/Problem3/MovieAction.java:29]: duplicated logic here too. 

suggesting:
```java
        int fee = lateFeePerDayInDollar * numOfDaysPastDue;
        return (numOfDaysPastDue >= 5) ? 2 * fee : fee;
```

File: src/test/java/Problem2Test.java [hunk 5]
```diff
@@ -7,13 +7,20 @@
...
+
         };
 
         int[][] expects = {
-                {0, 1, 2, 4, 6}
+                {0, 1, 2, 4, 6},
                 // add more test cases
+                {0, 1, 2, 3, 4, 6},
+                {0, 1, 2, 3, 4, 5, 6},
+                {0, 1}
```
Comment (pdgetrf) [inline src/test/java/Problem2Test.java:23]: great test. let's add more cases for the edge cases, such as null array and empty array.

**Review by pdgetrf — APPROVED** [review]
LGTM
